File: example/node/lib/rpc.js
var utils = require('./utils');

var path = require('path');
const protoLoader = require('@grpc/proto-loader');
const grpc = require('@grpc/grpc-js');

const protoPath = path.join("node_modules","nervatura","bin","proto","store.proto")

//const sslCreds = grpc.credentials.createSsl(sslCert);
const insecureCreds = grpc.credentials.createInsecure()

const pb = protoLoader.loadSync(protoPath, {
  keepCase: true, longs: Number, enums: String, defaults: false, oneofs: true
});
const nervatura = grpc.loadPackageDefinition(pb).nervatura

function rpcClient(ckey, meta, options, callback){
  const client = new nervatura.API(`localhost:${process.env.NT_GRPC_PORT}`, insecureCreds);
  client[ckey](options, meta, function(err, data){
    if(err && err.code === 14){
      client[ckey](options, meta, function(err, data){
        callback(err, data)
      })
    } else {
      callback(err, data)
    }
  });
}

exports.Database = function(options, callback) {
  var meta = new grpc.Metadata();
  meta.add('x-api-key', process.env.NT_API_KEY);
  rpcClient("Database", meta, options, function(err, data){
    if(err !== null){
      return callback(err.details, null)
    }
    var results = []
    if(data && data.data){
      results = utils.CheckJson(data.data)
    }
    callback(null, results)
  })
}

exports.CustomerUpdate = function(token, options, callback) {
  var meta = new grpc.Metadata();
  meta.add('x-api-key', process.env.NT_API_KEY);
  if(token){
    meta.add('Authorization', `Bearer ${token}`);
  }
  rpcClient("CustomerUpdate", meta, options, function(err, data){
    if(err !== null){
      return callback(err.details, null)
    }
    callback(null, data)
  })
}

exports.CustomerGet = function(token, options, callback) {
  var meta = new grpc.Metadata();
  meta.add('x-api-key', process.env.NT_API_KEY);
  if(token){
    meta.add('Authorization', `Bearer ${token}`);
  }
  rpcClient("CustomerGet", meta, options, function(err, data){
    if(err !== null){
      return callback(err.details, null)
    }
    callback(null, data)
  })
}

exports.CustomerQuery = function(token, options, callback) {
  var meta = new grpc.Metadata();
  meta.add('x-api-key', process.env.NT_API_KEY);
  if(token){
    meta.add('Authorization', `Bearer ${token}`);
  }
  rpcClient("CustomerQuery", meta, options, function(err, data){
    if(err !== null){
      return callback(err.details, null)
    }
    callback(null, data.data)
  })
}

exports.Delete = function(token, options, callback) {
  var meta = new grpc.Metadata();
  meta.add('x-api-key', process.env.NT_API_KEY);
  if(token){
    meta.add('Authorization', `Bearer ${token}`);
  }
  nt_model = nervatura.Model.type.value.filter(dtype => (dtype.name === options.model.toUpperCase()))
  if (nt_model.length > 0){
    options.model = nt_model[0].name
  }
  rpcClient("Delete", meta, options, function(err, data){
    if(err !== null){
      return callback(err.details, null)
    }
    callback(null, {})
  })
}

exports.Function = function(token, options, callback) {
  var meta = new grpc.Metadata();
  meta.add('x-api-key', process.env.NT_API_KEY);
  if(token){
    meta.add('Authorization', `Bearer ${token}`);
  }
  rpcClient("Function", meta, options, function(err, data){
    if(err !== null){
      return callback(err.details, null)
    }
    var results = {};
    if(data && data.data){
      results = utils.CheckJson(data.data)
    }
    callback(null, results)
  })
}

exports.View = function(token, options, callback) {
  var meta = new grpc.Metadata();
  meta.add('x-api-key', process.env.NT_API_KEY);
  if(token){
    meta.add('Authorization', `Bearer ${token}`);
  }
  rpcClient("View", meta, options, function(err, data){
    if(err !== null){
      return callback(err.details, null)
    }
    var results = [];
    if(data && data.data){
      results = utils.CheckJson(data.data)
    }
    callback(null, results)
  })
}

File: example/python/lib/rpc.py
import os
import grpc

from nervatura import store_pb2 as pb
from nervatura import store_pb2_grpc as api

import lib.utils as utils

def client():
  channel = grpc.insecure_channel("localhost:"+os.getenv("NT_GRPC_PORT"))
  return api.APIStub(channel)

def metadata(token):
  if token != "":
    return [("authorization", "Bearer "+token)]
  else:
    return [("x-api-key", os.getenv("NT_API_KEY"))]

def Database(options):
  try:
    response = client().Database(pb.RequestDatabase(
      alias=options["alias"],
      demo=options["demo"]
    ), metadata=metadata(""))
    result = utils.checkJson(response.data)
    return [result, None]
  except grpc.RpcError as err:
    return [None, err.details()]
  except Exception as e:
    return [None, str(e)]

def CustomerUpdate(token, data):
  try:
    response = client().CustomerUpdate(pb.Customer(
      code=data["code"],
      customer_name=data["customer_name"]
    ), metadata=metadata(token))
    return [response, None]
  except grpc.RpcError as err:
    return [None, err.details()]
  except Exception as e:
    return [None, str(e)]

def CustomerGet(token, options):
  try:
    response = client().CustomerGet(pb.RequestGet(
      code=options["code"]
    ), metadata=metadata(token))
    return [response, None]
  except grpc.RpcError as err:
    return [None, err.details()]
  except Exception as e:
    return [None, str(e)]

def CustomerQuery(token, filters):
  try:
    response = client().CustomerQuery(pb.RequestQuery(
      filters=[pb.RequestQueryFilter(field=field, value=value) for field, value in filters.items()]
    ), metadata=metadata(token))
    result = response.data
    return [result, None]
  except grpc.RpcError as err:
    return [None, err.details()]
  except Exception as e:
    return [None, str(e)]

def Delete(token, options):
  try:
    response = client().Delete(pb.RequestDelete(
      code=options["code"],
      model=pb.Model.Value(options["model"].upper())
    ), metadata=metadata(token))
    return [response, None]
  except grpc.RpcError as err:
    return [None, err.details()]
  except Exception as e:
    return [None, str(e)]

def Function(token, options):
  try:
    response = client().Function(pb.RequestFunction(
      function=options["name"],
      args=options["values"]
    ), metadata=metadata(token))
    result = utils.checkJson(response.data)
    return [result, None]
  except grpc.RpcError as err:
    return [None, err.details()]
  except Exception as e:
    return [None, str(e)]

def View(token, options):
  try:
    filters = [pb.RequestQueryFilter(field=filter["field"], value=filter["value"]) for filter in options["filters"]]
    response = client().View(pb.RequestView(
      name=pb.ViewName.Value(options["name"].upper()),
      filters=filters,
      limit=options["limit"],
    ), metadata=metadata(token))
    result = utils.checkJson(response.data)
    return [result, None]
  except grpc.RpcError as err:
    return [None, err.details()]
  except Exception as e:
    return [None, str(e)]